workers/worker-performance.worker.js

```diff
@@ -0,0 +1,128 @@
+
+importScripts("/resources/testharness.js");
+
+function verifyEntry (entry, name, type, duration, assertName) {
+    assert_equals(entry.name, name, assertName + " has the right name");
+    assert_equals(entry.entryType, type, assertName + " has the right type");
+    assert_equals(entry.duration, duration, assertName + "has the right duration");
+}
+
+var start;
+test(function testPerformanceNow () {
+    start = performance.now();
+}, "Can use performance.now in workers");
+
+test(function testPerformanceMark () {
+    performance.mark("mark1");
+     // Stall the minimum amount of time to ensure the marks are separate
+    var now = performance.now();
+    while (performance.now() == now) { }
+    performance.mark("mark2");
+}, "Can use performance.mark in workers");
+
+test(function testPerformanceMeasure () {
+    performance.measure("measure1", "mark1", "mark2");
+}, "Can use performance.measure in workers");
+
+test(function testPerformanceGetEntriesByName () {
+    var mark1s = performance.getEntriesByName("mark1");
+    assert_equals(mark1s.length, 1, "getEntriesByName gave correct number of entries");
+    verifyEntry(mark1s[0], "mark1", "mark", 0, "Entry got by name");
+}, "Can use performance.getEntriesByName in workers");
+
+var marks;
+var measures;
+test(function testPerformanceGetEntriesByType () {
+    marks = performance.getEntriesByType("mark");
+    assert_equals(marks.length, 2, "getEntriesByType gave correct number of entries");
+    verifyEntry(marks[0], "mark1", "mark", 0, "First mark entry");
+    verifyEntry(marks[1], "mark2", "mark", 0, "Second mark entry");
+    measures = performance.getEntriesByType("measure");
+    assert_equals(measures.length, 1, "getEntriesByType(\"measure\") gave correct number of entries");
+    verifyEntry(measures[0], "measure1", "measure", marks[1].startTime - marks[0].startTime, "Measure entry");
+}, "Can use performance.getEntriesByType in workers");
+
+test(function testPerformanceEntryOrder () {
+    assert_greater_than(marks[0].startTime, start, "First mark startTime is after a time before it");
+    assert_greater_than(marks[1].startTime, marks[0].startTime, "Second mark startTime is after first mark startTime");
+    assert_equals(measures[0].startTime, marks[0].startTime, "measure's startTime is the first mark's startTime");
+}, "Performance marks and measures seem to be working correctly in workers");
+
+test(function testPerformanceClearing () {
+    performance.clearMarks();
+    assert_equals(performance.getEntriesByType("mark").length, 0, "clearMarks cleared the marks");
+    performance.clearMeasures();
+    assert_equals(performance.getEntriesByType("measure").length, 0, "clearMeasures cleared the measures");
+}, "Can use clearMarks and clearMeasures in workers");
+
+test(function testPerformanceResourceTiming () {  // Resource timing
+    var start = performance.now();
+    var xhr = new XMLHttpRequest();
+     // Do a synchronous request and add a little artificial delay
```

Comment:
I don't recall the spec guaranteeing that when a synchronous XHR completes that a PerformanceResourceTiming entry is in the global performance buffer. That said, I think all existing resource timing tests relied on that behavior so all implementations have since met this pseudo-requirement. I think its fine to continue this, but it is something to be aware of.

Reply:
The spec says "All resources fetched by the current browsing [HTML5] or worker [WORKERS] context's MUST be included as PerformanceResourceTiming objects in the Performance Timeline of the relevant context." so XHR requests are expected to be included. See https://w3c.github.io/resource-timing/#resources-included

Reply:
My concern was not _if_ XHRs show up in the performance timeline it was _when_ they show in the performance timeline. I was thinking of https://github.com/w3c/resource-timing/issues/82 which the spec has a note about at the end of the processing model.

Re-reading the processing model, I would expect a sync XHR to have a Performance Entry added to the global performance timeline before JavaScript continues executing. So this test seems fine.